docs: Custom Identity Model Documentation (#453)
ory/docs: docs/docs/guides/customize-identity-model.mdx

The part highlighted below defines the identities `email` for the
email+password flow in the Ory Cloud. It also includes a method for
[`recovery`](https://www.ory.sh/kratos/docs/self-service/flows/account-recovery#recovery-link-method)
as well as
[`verification`](https://www.ory.sh/kratos/docs/self-service/flows/verify-email-account-activation/#verification-link-method).
Currently, only the `email` method is available (recovery/verification via a
link sent in an email).

```diff
    "traits": {
      "type": "object",
      "properties": {
+       "email": {
+         "type": "string",
+         "format": "email",
+         "title": "E-Mail",
+         "ory.sh/kratos": {
+           "credentials": {
+             "password": {
+               "identifier": true
+             }
+           },
            "recovery": {
              "via": "email"
            },
            "verification": {
              "via": "email"
            }
          }
        }
      },
      "required": [
        "username"
      ],
      "additionalProperties": false
    }
```

To add traits to the Identity Model, add them inside the `traits "properties"`

> ```json

Reviewer: Use json5 if you include comments

Author: ```suggestion
```json5
```
Do you mean this? 
It loses the color highlights if I use `json5` here. I think using `json` here, already renders it correctly.

with json5
<img width="968" alt="2" src="https://user-images.githubusercontent.com/55394283/128321235-c5018d71-948b-4607-914d-ad2805e41558.png">

with json
<img width="968" alt="1" src="https://user-images.githubusercontent.com/55394283/128321215-dfb603c8-010d-4ff9-83ff-a79267cd4c73.png">

Reviewer: I think we just need to add json5 as an alias in the docs config. It would fix the markdown preview in github and potentially any CI we introduce in the future to validate markdown

Author: I changed it now in https://github.com/ory/docs/pull/453/commits/1ecf575728dede34d18a9ff64db96c5ee6194aaf to not require comments anymore. 
I think it is better this way. But we should still add json5 as alias. 
If you can give me some pointers, I could not figure out how to add it.

"traits": {
"type": "object",
"properties": {
// your identities traits go here
       }
    }
```

for example the GitHub Handle as `string` :

```json
"traits": {
"type": "object",
"properties": {

>         "handle": {

Reviewer: Incorrect indentation

---
id: customize-identity-model

> title: Customize Identity Model

Reviewer: Sorry, I just saw this - while it is named "Identity Model" we really should change it to "Custom Identity Schema". We will change this in all the UIs etc. 

So please replace `Identity Model` with `Identity Schema`!

Author: I was going off this sentence in the identity concept docs:
`The Identity Model is a JSON Schema which describes your identity's (e.g. your customers / users / ...) fields.`

Do you think it would make sense to add the glossary from my other PR soon, for the different terms like Identity Model/Schema, Project Slug, etc: ?
https://github.com/ory/docs/blob/d0465fff31431b288ba9c1087a3757b7f160a6fc/docs/docs/concepts/self-service.mdx#glossary

I changed it, I hope it is clear in this document:
https://github.com/ory/docs/blob/a7438c42b2f0b837a9485d2a6d71c37ecc272953/docs/docs/concepts/identity.mdx

---
id: customize-identity-model
title: Customize Identity Model
---

Ory supports custom Identity Models. To learn more about the concept of Identity
Models at Ory, please head over to
[Identity Data Model, Custom Fields](../concepts/identity.mdx).

This guide will help you create custom Identity Models as well as provide an
example Identity Model.

## Setup Custom Identity Model

- Select the Ory Cloud Project, you want to change under
  [Projects/Overview](https://console.ory.sh/projects/list).

- Head over to the "Customize / General" tab (URL:
  https://console.ory.sh/projects/$YOURPROJECTID/settings) and review the
  currently active Identity Model.

- Select a preset model or an empty template as the basis for your new custom
  Identity Model.

- Check the `Customize Identity Schema` box to enable editing on the Identity
  Model.

- Add or remove traits from the Identity Model as desired, more info in the
  [Customize Fields](#customize-fields) section.

- When finished editing name your custom Identity Model. Input the name in the
  textbox between the checkbox and the schema.

- Press `Enter` or hit the `Update` button to save it, once you have made your

>   changes. Currently, it is not possible to edit an existing Identity Model, so

Reviewer 2: Maybe this needs some more details such as

```suggestion
  changes. Currently, it is not possible to edit an existing Identity Model on the same revision. The identity model can still be reused as a template for a following revision.
```

Then maybe have a section explaining what the revision would look like:

Revisions 1:
 - I create a schema with name "MyUsersv1"
 - Create the identity model and save it

Revisions 2:
 - Select the schema "MyUsersv1" as a template
 - Enter a new name (can technically be the same but there will be 2 of them then with different creation times)
 - Make the necessary adjustments to the identity model
 - Hit Save

Author: Great Idea! I will add such a section.

Currently, only the `email` method is available (recovery/verification via a link
sent in an email).

```diff
    "traits": {
      "type": "object",
      "properties": {
+       "username": {
+         "type": "string",
+         "title": "Username",
+         "ory.sh/kratos": {
+           "credentials": {
+             "password": {
+               "identifier": true
+             }
+           },
            "recovery": {
              "via": "email"
            },
            "verification": {
              "via": "email"
            }
          }
        }
      },
      "required": [
        "username"
      ],
      "additionalProperties": false
    }
```

To add traits to the Identity Model, add them inside the `traits "properties"`

```json
"traits": {
"type": "object",
"properties": {
// your identities traits go here
       }
    }
```

for example the GitHub Handle as `string` :

```json
"traits": {
"type": "object",
"properties": {
        "handle": {
          "type": "string",
          "title": "Your GitHub Handle"
        }
      }
    }
```

> or a mobile phone number as `integer`:

Reviewer 2: phone number as an integer won't be a good user experience.
It will create a input field with a "increase/decrease" button on the side - which looks bad.

Author: True, I noticed that as well. 
I think `age` would probably a good example for now.

`number` for integral or floating-point numbers. If you want to know more about
these types, please refer to the
[json-schema documentation](https://json-schema.org/understanding-json-schema/reference/type.html).

The `title` of each field is what the user as description or sample input. After
adding the above examples the sign-up screen would look like so:

![Example Identity Model with GitHub and Phone Number Fields](../images/identity-model/example-identity0.png)

### Required traits

To define certain traits as required, add them to the `required` array by name:

```json
  "properties": {
    "traits": {
      "type": "object",
      "properties": {
// your identities traits go here
       },
      "required": [
/add required fields to this array
        "email"
      ],
    }
  }
```

### Additional Properties

The
[`additionalProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties)
keyword is used to control the handling of properties whose names are not listed
in the properties keyword. This usually should be set to false.

```json
      "additionalProperties": false
```

> 

Reviewer 2: As far as I can remember this wouldn't make a difference currently to the managed UI - only to custom UI's.

Author: ```suggestion
The
[`additionalProperties`](https://json-schema.org/understanding-json-schema/reference/object.html#additional-properties)
keyword is used to control the handling of properties whose names are not listed
in the properties keyword. This currently has no effect and should be set to false.

```json
      "additionalProperties": false
```

Maybe like so? 

I wanted to include it, since not mentioning it at all would also be confusing IMO.

## Example Identity Model

The following Identity Model includes first/last and nickname, as well as number
fields for the phone number and birthday. There are also two true/false fields

> for specifying the newsletter subscription and enterprise status.

Reviewer 2: integer for phone number isn't ideal

Author: I will change it to `age` 🙂 

Do you think it makes sense to include a notice that phone numbers are not really supported right now?
